### PR TITLE
Represent GVK information as labels

### LINKED ISSUE #1847: CRD metrics are extremely brittle
**What happened**:

CRD metrics are currently extremely brittle due to the way it is implemented. The name of the CRD is interpolated into the name of the metric which makes it impossible to aggregate a CRD across different versions (or at least very very annoying to do so). 

**What you expected to happen**:

The name, version, group of the CRD should be a value in a metric label.

**How to reproduce it (as minimally and precisely as possible)**:

https://github.com/kubernetes/kube-state-metrics/blob/master/docs/customresourcestate-metrics.md

### PR DESCRIPTION
**Pull request by rexagod**
**What this PR does / why we need it**: Represent GVK information as labels in the metrics, instead of appending them to the metric name itself. This would allow users to aggregate varying GVKs of a CR under the same metric, making operations much more easier.

**How does this change affect the cardinality of KSM**: Increases, or stays the same, depending on the cardinality of the group(s), version(s), or kind(s) of a CR present in the cluster at the given time.

**Which issue(s) this PR fixes**: Fixes #1847

### CODE REVIEW

File: pkg/customresourcestate/registry_factory_test.go [hunk 1]
```diff
@@ -324,7 +324,7 @@ func Test_fullName(t *testing.T) {
 				resource: r(nil),
 				f:        count,
 			},
-			want: "kube_apps_v1_Deployment_count",
+			want: "count",
```
Comment (logicalhan) [inline pkg/customresourcestate/registry_factory_test.go:327]: should we prefix these with `crd_`?

**Review by logicalhan — COMMENTED** [review]
This looks awesome by the way. But I have some reservations about the unqualified metric names, because they seem way too generic.

File: pkg/customresourcestate/registry_factory_test.go (hunk 1 above)
Comment (mrueg) [reply pkg/customresourcestate/registry_factory_test.go:327]: I'd prefer kube_crd

File: docs/customresourcestate-metrics.md [hunk 2]
```diff
@@ -287,7 +289,7 @@ spec:
 
 Produces:
 ```prometheus
-myteam_foos_uptime 43.21
+myteam_foos_uptime{group="myteam.io", kind="Foo", version="v1"} 43.21
```
Comment (logicalhan) [inline docs/customresourcestate-metrics.md:292]:
```suggestion
uptime{group="myteam.io", kind="Foo", version="v1"} 43.21
```

Comment (logicalhan) [reply docs/customresourcestate-metrics.md:292]: Would you mind changing the prefix to "kube_crd" as the other comment suggested?

Comment (logicalhan) [reply docs/customresourcestate-metrics.md:292]: https://github.com/kubernetes/kube-state-metrics/pull/1850#discussion_r991572154

File: pkg/customresourcestate/config.go [hunk 3]
```diff
@@ -63,17 +62,11 @@ type Resource struct {
...
 func (r Resource) GetMetricNamePrefix() string {
-	if r.MetricNamePrefix == nil {
-		return strings.NewReplacer(
-			"/", "_",
-			".", "_",
-			"-", "_",
-		).Replace(fmt.Sprintf("kube_%s_%s_%s", r.GroupVersionKind.Group, r.GroupVersionKind.Version, r.GroupVersionKind.Kind))
+	p := r.MetricNamePrefix
+	if p == nil {
+		return "kube"
```
Comment (logicalhan) [inline pkg/customresourcestate/config.go:67]:
```suggestion
		return "kube_crd"
```

**Comment by rexagod** [comment]
@logicalhan Addressed the reviews, PTAL.

File: pkg/customresourcestate/registry_factory_test.go [hunk 4]
```diff
@@ -324,7 +324,7 @@ func Test_fullName(t *testing.T) {
 				resource: r(nil),
 				f:        count,
 			},
-			want: "kube_apps_v1_Deployment_count",
+			want: "kube_count",
```
Comment (logicalhan) [inline pkg/customresourcestate/registry_factory_test.go:327]:
```suggestion
			want: "kube_crd_count",
```

File: docs/customresourcestate-metrics.md (hunk 2 above)
Comment (rexagod) [reply docs/customresourcestate-metrics.md:292]: This metric would be unchanged as the configuration specified a `metricNamePrefix: myteam_foos`.

https://github.com/kubernetes/kube-state-metrics/blob/633cbf49d0ecf607c716860b21e03c4afd596b8c/docs/customresourcestate-metrics.md?plain=1#L279-L293

File: docs/customresourcestate-metrics.md [hunk 5]
```diff
@@ -305,6 +305,11 @@ spec:
           ...
 ```
 
+Produces:
+```prometheus
+uptime{group="myteam.io", kind="Foo", version="v1"} 43.21
+```
+
```
Comment (rexagod) [inline docs/customresourcestate-metrics.md:312]: @logicalhan JFYI I added a doc snippet here to make things a bit more explicit to the reader.

Comment (logicalhan) [reply docs/customresourcestate-metrics.md:312]: sgtm

**Pull request merged**